---
output: 
    html_document:
        theme: spacelab
        toc: true
        toc_float: true
        code_download: true
toc_collapsed: true
params:
  pactName: "PACT-YY-##"
  userName: "Jonathan Serrano"
  workDir: "/Volumes/CBioinformatics/jonathan/pact/consensus"
title: "Consensus `r params$pactName`"
author: "Script Vers: 1.1.1\n\nGenerated by: `r paste0(Sys.info()[['user']])`"
date: "`r format(Sys.time(), '%d %B, %Y')`"
editor_options: 
  chunk_output_type: console
---

```{r setup, include=FALSE}
gb <- globalenv(); assign("gb", gb)
devtools::source_url("https://raw.githubusercontent.com/NYU-Molecular-Pathology/Methylation/main/PACT_scripts/KnitScripts.R")
gb$bamURL <- gb$GetBamUrl(params)
gb$loadHtmlTag()
```

```{css, echo=FALSE, code=gb$css_code}
```
<h4>**Login with Kerberos to view BAM files in**: [`r gb$bamURL`](`r gb$bamURL`)</h4><br>
```{r, label='Input Parameters', opts.label='input_opts'}
philipsFtp <- "/Volumes/molecular/Molecular/Philips_SFTP" # Philips datadump path
methDir <- "/Volumes/molecular/Molecular/MethylationClassifier/CNV_PNG"
critialMnts <- c("/Volumes/CBioinformatics/Methylation/", "/Volumes/molecular/")
# saveRDS(params, file="params.rds") # Debug params <- readRDS("params.rds")
gb$CheckMntDirs(critialMnts, params)
gb$CheckFiExist(params$pactName, philipsFtp)
gb$CopyCnvPngs(params)
gb$MoveFacets()
gb$GetMethCnv(params, methDir)
gb$CheckHotspots(params)
```

```{r, label='Table Output', opts.label='tab_opts'}
gb$LoopSampleTabs(params) # Debug: Run first chunk then View(gb$LoopSampleTabs)
```
